### PR TITLE
Edit pass for ECS Readme.md

### PR DESCRIPTION
**Pull request by karenzone**
**Goals for structure:**
- Present info in logical order 
- Reduce repetition
- Add navigation tools (mini-TOC) so that users can:
     - Get a sense of what info is in the document
     - Jump to what they want to see

**General comments/questions:**

- If all fields are (more or less) user-defined, it’s confusing to call out just a few as user-defined.
- “TBD” notation calls out other questions or areas to follow up on

**Suggestion for enhancements:**
-   More examples! Flesh out examples to illustrate suggested use for field, and help direct users down the right path.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,83 +2,119 @@
...
 
-| Field  | Description  | Type  | Multi Field  | Example  |
-|---|---|---|---|---|
-| <a name="@timestamp"></a>`@timestamp`  | Timestamp when the event was created.<br/>For log events this is expected to be when the event was generated and not when it was read.<br/>Timestamp is a required field and must exist in all events.  | date  |   | `2016-05-23T08:05:34.853Z`  |
-| <a name="tags"></a>`tags`  | Tags is a list of keywords which are used to tag each event.  | keyword  |   | `["production", "env2"]`  |
-| <a name="labels"></a>`labels`  | Labels is an object which contains key/value pairs.<br/>Labels can be used to add additional meta information to events. Label should not contain nested objects and all values are stored as keyword.<br/>An example usage is the docker and k8s labels.  | object  |   | `{key1: value1, key2: value2}`  |
-| <a name="message"></a>`message`  | For log events the message field contains the log message.<br/>In other use cases the message field can be used to concatenate together different values which are then freely searchable. Or if multiple messages exist they can be combined here into one message.  | text  |   | `Hello World`  |
+**Guidelines**
```
Comment (ruflin) [inline README.md]: I like this part on the top here.

File: README.md [hunk 2]
```diff
@@ -2,83 +2,119 @@
...
+
+* The document MUST have the `@timestamp` field.
+* The [data type](https://www.elastic.co/guide/en/elasticsearch/reference/6.2/mapping-types.html) defined for an ECS field MUST be used.
+* It SHOULD have the field `event.version` to define which version of ECS it uses.
+* As many fields as possible should be mapped to ECS.
 
+**Writing fields**
```
Comment (ruflin) [inline README.md]: These are rules for ECS field names. To know these rules is only required when adding new fields. Should we move this to the contribution guide?

File: README.md [hunk 3]
```diff
@@ -2,83 +2,119 @@
...
 
-## <a name="agent"></a> Agent fields
+* All fields must be lower case
+* No special characters except `_`
+* Use underscore to combine words
 
-The agent fields contains the data about the agent/client/shipper that created the event.
+**Naming Fields**
```
Comment (ruflin) [inline README.md]: The same is true for this part: Only relevant when contributing fields. I wonder if we should move it to the bottom or contributing guide.

File: README.md [hunk 4]
```diff
@@ -2,83 +2,119 @@
...
+* *Present tense.* Use present tense unless field describes historical information.
+* *Singular or plural.* Use singular and plural names properly to reflect the field content. For example, use `requests_per_sec` rather than `request_per_sec`.
+* *General to specific.* Organise the prefixes from general to specific to allow grouping fields into objects with a prefix like `host.*`.
+* *Avoid repetition.* Avoid stuttering of words. If part of the field name is already in the prefix, do not repeat it. Example: `host.host_ip` should be `host.ip`.
+* *Use prefixes.* Fields must be prefixed except for the base fields. For example all `host` fields are prefixed with `host.`. See `dot` notation in FAQ for more details.
+* Do not use abbreviations (few exceptions like `ip` exist)
+
+## <a name="fields"></a>Fields
+
+[Base fields](#base)
```
Comment (ruflin) [inline README.md]: I need to adjust the script to create this list.

File: README.md [hunk 5]
```diff
@@ -2,83 +2,119 @@
...
+| [Process fields](#process)
+| [Service fields](#service)
+| [Source fields](#source)
+| [URL fields](#url)
+| [User fields](#user)
+| [User agent fields](#user_agent)
+
+### <a name="base"></a> Base fields
+
+The base set contains all fields at the top level without a namespace. These are fields which are common across all types of events.
```
Comment (ruflin) [inline README.md]: Let's not use `namespace` but better prefix. This caused quite some confusion in the past so I tried to remove the word `namespace` everywhere.

File: README.md [hunk 6]
```diff
@@ -2,83 +2,119 @@
...
 
 
 | Field  | Description  | Type  | Multi Field  | Example  |
```
Comment (ruflin) [inline README.md]: +1 on putting these fields on top. I must modify the script for this.

File: README.md [hunk 7]
```diff
@@ -89,10 +125,11 @@ In case Metricbeat is running on an EC2 host and fetches data from its host, the
 | <a name="cloud.instance.name"></a>`cloud.instance.name`  | Instance name of the host machine.  | keyword  |   |   |
 | <a name="cloud.machine.type"></a>`cloud.machine.type`  | Machine type of the host machine.  | keyword  |   | `t2.medium`  |
 
+Examples: If Metricbeat is running on an EC2 host and fetches data from its host, the cloud fields are expected to contain data about the host machine. If Metricbeat is running on a remote machine outside of the cloud and fetches data from a service running in the cloud, cloud fields are expected to reflect the data from the cloud machine which is running the service.
 
-## <a name="container"></a> Container fields
+### <a name="container"></a> Container fields
```
Comment (ruflin) [inline README.md]: Need to adjust the script to add one more # here.

File: README.md [hunk 8]
```diff
@@ -105,45 +142,41 @@ Container fields are used for meta information about the specific container the
...
+| <a name="device.timezone.offset.sec"></a>`device.timezone.offset.sec`  | Timezone offset of the host in seconds.<br/>Number of seconds relative to UTC.<br/> Example: If the offset is -01:30, the value will be -5400.  | long  |   | `-5400`  |
+| <a name="device.type"></a>`device.type`  | The type of the device that the data is coming from.<br/>There is no predefined list of device types. <br/>Examples: `endpoint`, `firewall`, `ids`, `ips`, `proxy`.  | keyword  |   | `firewall`  |
 
-## <a name="error"></a> Error fields
 
-Error namespace
+### <a name="error"></a> Error fields
 
-This can be used to represent all kinds of errors. It can be for errors that happen while fetching events or if the event itself contains an error.
+The `error` namespace can represent errors of any kind. Use it for errors that happen while fetching events or if the event itself contains an error.
```
Comment (ruflin) [inline README.md]: Let's not use `namespace`, see note above.

File: README.md [hunk 9]
```diff
@@ -153,93 +186,92 @@ This can be used to represent all kinds of errors. It can be for errors that hap
 | <a name="error.code"></a>`error.code`  | Error code describing the error.  | keyword  |   |   |
 
 
-## <a name="event"></a> Event fields
+### <a name="event"></a> Event fields
 
-The event fields are used for context information about the data itself.
+The event fields are carry context information about the data itself.
```
Comment (ruflin) [inline README.md]: The event fields carry context?

File: README.md [hunk 10]
```diff
@@ -153,93 +186,92 @@ This can be used to represent all kinds of errors. It can be for errors that hap
...
-| <a name="event.severity"></a>`event.severity`  | Severity describes the severity of the event. What the different severity values mean can very different between use cases. It's up to the implementer to make sure severities are consistent across events.  | long  |   | `7`  |
-| <a name="event.raw"></a>`event.raw`  | Raw text message of entire event to be used to demonstrate log integrity.  | keyword  |   | `Sep 19 08:26:10 host CEF:0&#124;Security&#124; threatmanager&#124;1.0&#124;100&#124; worm successfully stopped&#124;10&#124;src=10.0.0.1 dst=2.1.2.2spt=1232`  |
+| <a name="event.category"></a>`event.category`  | Event category.<br/>TBD. This can be a user defined category.  | keyword  |   | `metrics`  |
+| <a name="event.type"></a>`event.type`  | A type given to this kind of event which can be used for grouping.<br/>TBD. This is normally defined by the user.  | keyword  |   | `nginx-stats-metrics`  |
+| <a name="event.module"></a>`event.module`  | Name of the Beats or Logstash module that is the source of the data.  | keyword  |   | `mysql`  |
```
Comment (ruflin) [inline README.md]: I would not limit this to `Beats` or `Logstash` here but rather use the two as examples.

File: README.md [hunk 11]
```diff
@@ -153,93 +186,92 @@ This can be used to represent all kinds of errors. It can be for errors that hap
...
-| <a name="event.dataset"></a>`event.dataset`  | Name of the dataset.<br/>The concept of a `dataset` (fileset / metricset) is used in Beats as a subset of modules. It contains the information which is currently stored in metricset.name and metricset.module or fileset.name.  | keyword  |   | `stats`  |
-| <a name="event.severity"></a>`event.severity`  | Severity describes the severity of the event. What the different severity values mean can very different between use cases. It's up to the implementer to make sure severities are consistent across events.  | long  |   | `7`  |
-| <a name="event.raw"></a>`event.raw`  | Raw text message of entire event to be used to demonstrate log integrity.  | keyword  |   | `Sep 19 08:26:10 host CEF:0&#124;Security&#124; threatmanager&#124;1.0&#124;100&#124; worm successfully stopped&#124;10&#124;src=10.0.0.1 dst=2.1.2.2spt=1232`  |
+| <a name="event.category"></a>`event.category`  | Event category.<br/>TBD. This can be a user defined category.  | keyword  |   | `metrics`  |
```
Comment (ruflin) [inline README.md]: Lets move the discussion about type and category to a separate PR afterwards.

File: README.md [hunk 12]
```diff
@@ -153,93 +186,92 @@ This can be used to represent all kinds of errors. It can be for errors that hap
...
+| <a name="event.created"></a>`event.created`  | Date the event was created.<br/>**TBD.** @timestamp and event.created both use "Time the event was created" as the description. Can we further clarify by saying WHO is doing the creating of each?<br/><br/>This field is distinct from @timestamp, even though the values may be the same in some cases.<br/>Examples: For logs, `@timestamp` contains the timestamp carried in the log line, and `event.created` contains the time when the log line is read. For package captures, `@timestamp` contains the timestamp extracted from the network package, and `event.created` contains the time the event was created. </br>If the two timestamps are identical, use [@timestamp](#timestamp). | date  |   |   |
```
Comment (ruflin) [inline README.md]: +1 on making the destinction very clear. I hoped the examples help here.

File: README.md [hunk 13]
```diff
@@ -153,93 +186,92 @@ This can be used to represent all kinds of errors. It can be for errors that hap
...
+| <a name="file.device"></a>`file.device`  | Device that is the source of the file.  | keyword  |   |   |
+| <a name="file.inode"></a>`file.inode`  | Inode representing the file in the filesystem.  | keyword  |   |   |
+| <a name="file.uid"></a>`file.uid`  | User ID (UID) or security identifier (SID) of the file owner.  | keyword  |   |   |
+| <a name="file.owner"></a>`file.owner`  | File owner's username.  | keyword  |   |   |
+| <a name="file.gid"></a>`file.gid`  | Primary group ID (GID) of the file.  | keyword  |   |   |
+| <a name="file.group"></a>`file.group`  | Primary group name of the file.  | keyword  |   |   |
+| <a name="file.mode"></a>`file.mode`  | Mode of the file in octal representation.  | keyword  |   | `416`  |
+| <a name="file.size"></a>`file.size`  | File size in bytes (field is only added when `type` is `file`).  | long  |   |   |
+| <a name="file.mtime"></a>`file.mtime`  | Last time file content was modified.  | date  |   |   |
+| <a name="file.ctime"></a>`file.ctime`  | Last time  file metadata changed.  | date  |   |   |
```
Comment (ruflin) [inline README.md]: Added space?

File: README.md [hunk 14]
```diff
@@ -153,93 +186,92 @@ This can be used to represent all kinds of errors. It can be for errors that hap
...
+| <a name="host.id"></a>`host.id`  | Unique host id.<br/>As hostname is not always unique, this often can be configured by the user.<br/>TBD: "This" = What speficially? <br/>Example: The current usage of `beat.name`.  | keyword  |   |   |
 | <a name="host.ip"></a>`host.ip`  | Host ip address.  | ip  |   |   |
 | <a name="host.mac"></a>`host.mac`  | Host mac address.  | keyword  |   |   |
-| <a name="host.type"></a>`host.type`  | This is the type of the host.<br/>For Cloud providers this can be the machine type like `t2.medium`. Or it vm, container for example or something user defined.  | keyword  |   |   |
-| <a name="host.os.platform"></a>`host.os.platform`  | Operating system platform (e.g. centos, ubuntu, windows).  | keyword  |   | `darwin`  |
+| <a name="host.type"></a>`host.type`  | TBD</br>Type of host.<br/>For Cloud providers this can be the machine type, such as `t2.medium`. Or it vm, container for example or something user defined.  | keyword  |   |   |
```
Comment (ruflin) [inline README.md]: Let's move this also to a separate PR for discussion.

File: README.md [hunk 15]
```diff
@@ -263,156 +295,142 @@ Fields which are specific to log events.
...
 
 
 | Field  | Description  | Type  | Multi Field  | Example  |
 |---|---|---|---|---|
-| <a name="service.id"></a>`service.id`  | Unique identifier of the running service.<br/>This id should uniquely identify this service. This makes it possible to correlate logs and metrics for one specific service. For example in case of issues with one redis instance, it's possible to filter on the id to see metrics and logs for this single instance.  | keyword  |   | `d37e5ebfe0ae6c4972dbe9f0174a1637bb8247f6`  |
-| <a name="service.name"></a>`service.name`  | Name of the service data is collected from.<br/>The name can be used to group logs and metrics together from one service and correlate them.  | keyword  |   | `elasticsearch`  |
+| <a name="service.id"></a>`service.id`  | Unique identifier of the service that is running.<br/>The id must uniquely identify this service. This makes it possible to correlate logs and metrics for one specific service.</br> Example: If one redis instance is experiencing problems, you can filter on the id to see metrics and logs for that single instance.  | keyword  |   | `d37e5ebfe0ae6c4972dbe9f0174a1637bb8247f6`  |
```
Comment (ruflin) [inline README.md]: `Redis` should probably be upper case here.

File: README.md [hunk 16]
```diff
@@ -2,83 +2,119 @@
...
 For contributions please read the [Contributing Guide](CONTRIBUTING.md).
 
-The current version of ECS is `0.1.0`.
+<a name="ecs-version"></a>The current version of ECS is `0.1.0`.
 
+## In this readme
+
+* [Implementing ECS](#implementing-ecs)
 * [Fields](#fields)
+  * [Base fields](#base)
```
[the remark this answers is not in the document]
Comment (karenzone) [reply README.md]: Is the minor change in structure worth the effort to modify the script?  I like the  fields nested in a mini-TOC, but it might not be worth the time/effort to change the script.

File: README.md [hunk 17]
```diff
@@ -2,83 +2,119 @@
 
 # Elastic Common Schema (ECS)
 
-The Elastic Common Schema (ECS) is used to provide a common data model when
-ingesting data into Elasticsearch. Having a common schema allows you correlate
-data from sources like logs and metrics or IT operations
-analytics and security analytics.
+The Elastic Common Schema (ECS) provides a common data model for
```
[the remark this answers is not in the document]
Comment (karenzone) [reply README.md]: Possible alternatives to "common data model":  pattern? structure? syntax? framework? 
Are any of those a better alternative?

File: README.md (hunk 4 above)
Comment (karenzone) [reply README.md]: @ruflin  Is this helpful enough to warrant the effort?  The intention is  a compact, clickable list to reduce need for scrolling.

File: README.md (hunk 2 above)
Comment (karenzone) [reply README.md]: I trust your judgement.  I like having the guidelines here because it summarizes the approach for naming fields.  In the benefits, we talk about improved ability to remember commonly used names and improved ability to remember field names that have been forgotten. IMO these guidelines contribute to the goals/benefits.

File: README.md (hunk 5 above)
Comment (karenzone) [reply README.md]: Suggestions for replacement?

File: README.md (hunk 12 above)
Comment (karenzone) [reply README.md]: Examples helped for sure!  Adding a few more (or slightly different) words describing "who" or "when" would help even more to differentiate. I tried "Data/time the event originated" for @timestamp. Is that accurate?

File: README.md (hunk 14 above)
Comment (karenzone) [reply README.md]: This is flagged "TBD" because I couldn't figure out what we're trying to say here:
`Or it vm, container for example or something user defined.`
  I'll take a crack at it. If I guess incorrectly, we can explore in another PR.

File: README.md (hunk 3 above)
Comment (karenzone) [reply README.md]: Helps with remembering/guessing?

File: README.md (hunk 9 above)
Comment (karenzone) [reply README.md]: Tried different wording.  Can change back to original if preferred.

File: README.md (hunk 11 above)
Comment (karenzone) [reply README.md]: I gave it a shot. If what I proposed doesn't work, we'll move to a separate PR.

**Comment by dedemorton** [comment]
>Will that treatment work in a generated MD file--maybe if I provide more details in the link? 

@karenzone I don't know. I haven't worked with this file and don't know markdown very well. Maybe ask @ruflin can advise?

@ruflin The script should also add a comment so we know the file is generated. :-) 

@karenzone Sounds like we just need to get these changes into the source docs. Let me know if you need help figuring out what to update in the docs source. Thanks SO much for doing this!!

**Comment by karenzone** [comment]
This PR turned into a meta-issue that contains proposals and comments to be implemented in individual source files (#22  and #24)
